### PR TITLE
[logger] rework logging facility

### PR DESCRIPTION
**Pull request by ufoot**

### CODE REVIEW

File: lib/ddtrace/tracer.rb [hunk 1]
```diff
@@ -21,12 +22,26 @@ class Tracer
...
-        @logger = Logger.new(STDOUT)
+        @logger = Datadog::Logger.new(STDOUT)
         @logger.level = Logger::WARN
       end
       @logger
     end
 
+    # Override the default logger with a custom one.
+    def self.log=(logger)
+      return nil unless logger
```
Comment (palazzem) [inline lib/ddtrace/tracer.rb]: should we return `nil` here? just a `return` isn't enough?

File: lib/ddtrace/tracer.rb [hunk 2]
```diff
@@ -21,12 +22,26 @@ class Tracer
...
+      return nil unless logger.respond_to? :methods
+      return nil unless logger.respond_to? :error
+      if logger.respond_to? :methods
+        unimplemented = Logger.new(STDOUT).methods - logger.methods
+        unless unimplemented.empty?
+          logger.error("logger #{logger} does not implement #{unimplemented}")
```
Comment (palazzem) [inline lib/ddtrace/tracer.rb]: do we need to `return` here? I mean if the new logger doesn't implement the `Logger` "interface", is it right to set it anyway?

File: test/logger_test.rb [hunk 3]
```diff
@@ -25,4 +27,84 @@ def test_tracer_set_debug_mode
...
+    assert_equal(true, Datadog::Tracer.log.fatal?)
+
+    Datadog::Tracer.log.debug('never to be seen')
+    Datadog::Tracer.log.warn('careful here')
+    Datadog::Tracer.log.error() { 'this does not work' }
+    Datadog::Tracer.log.error('mmm') { 'neither does this' }
+
+    lines = buf.string.lines
+
+    # Test below iterates on lines, this is required for Ruby 1.9 backward compatibility.
```
Comment (palazzem) [inline test/logger_test.rb]: 👍

File: lib/ddtrace/tracer.rb [hunk 4]
```diff
@@ -21,12 +22,27 @@ class Tracer
...
       end
       @logger
     end
 
+    # Override the default logger with a custom one.
+    def self.log=(logger)
+      return unless logger
+      return unless logger.respond_to? :methods
+      return unless logger.respond_to? :error
+      if logger.respond_to? :methods
```
Comment (Ferdy89) [inline lib/ddtrace/tracer.rb]: I think this line is redundant after
```ruby
return unless logger.respond_to? :methods
```

**Pull request merged**